### DISCUSSION TITLE
Decrypt Message

### DISCUSSION THREAD

**Discussion by esthicodes** [Estheryu991/SwissGerman_Dictionary#160] [Q&A]
<h2 style="font-style: normal; font-variant-caps: normal; orphans: auto; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none; box-sizing: border-box; width: auto; margin-top: 5px; margin-bottom: 20px; float: none; font-family: serif; color: rgb(101, 99, 99); font-size: 27px; line-height: 36px; font-weight: 700; text-align: left; letter-spacing: 1px; background-color: rgb(255, 255, 255);">Decrypt Message</h2><p style="font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none; box-sizing: border-box; display: block; margin-bottom: 10px; padding: 0px; color: rgb(77, 77, 77); font-size: 18px; line-height: 1.5; text-align: left; word-wrap: break-word; font-family: serif; background-color: rgb(255, 255, 255);">An infamous gang of cyber criminals named “The Gray Cyber Mob”, which is behind many hacking attacks and drug trafficking, has recently become a target for the FBI. After intercepting some of their messages, which looked like complete nonsense, the agency learned that they indeed encrypt their messages, and studied their method of encryption.</p><p style="font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none; box-sizing: border-box; display: block; margin-bottom: 10px; padding: 0px; color: rgb(77, 77, 77); font-size: 18px; line-height: 1.5; text-align: left; word-wrap: break-word; font-family: serif; background-color: rgb(255, 255, 255);">Their messages consist of lowercase latin letters only, and every word is encrypted separately as follows:</p><p style="font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none; box-sizing: border-box; display: block; margin-bottom: 10px; padding: 0px; color: rgb(77, 77, 77); font-size: 18px; line-height: 1.5; text-align: left; word-wrap: break-word; font-family: serif; background-color: rgb(255, 255, 255);">Convert every letter to its<span class="Apple-converted-space"> </span><a href="http://asciitable.com/" title="ASCII value" target="_blank" style="box-sizing: border-box; text-decoration: underline; color: rgb(156, 109, 226);">ASCII value</a>. Add<span class="Apple-converted-space"> </span><code style="box-sizing: border-box; font-family: &quot;Droid Sans Mono&quot;, monospace; font-weight: inherit; background-color: rgba(0, 0, 0, 0.04); border-radius: 3px; font-size: 15px; line-height: 1.5; padding: 0.1em 0.3em;">1</code><span class="Apple-converted-space"> </span>to the first letter, and then for every letter from the second one to the last one, add the value of the previous letter. Subtract<span class="Apple-converted-space"> </span><code style="box-sizing: border-box; font-family: &quot;Droid Sans Mono&quot;, monospace; font-weight: inherit; background-color: rgba(0, 0, 0, 0.04); border-radius: 3px; font-size: 15px; line-height: 1.5; padding: 0.1em 0.3em;">26</code><span class="Apple-converted-space"> </span>from every letter until it is in the range of lowercase letters<span class="Apple-converted-space"> </span><code style="box-sizing: border-box; font-family: &quot;Droid Sans Mono&quot;, monospace; font-weight: inherit; background-color: rgba(0, 0, 0, 0.04); border-radius: 3px; font-size: 15px; line-height: 1.5; padding: 0.1em 0.3em;">a-z</code><span class="Apple-converted-space"> </span>in ASCII. Convert the values back to letters.</p><p style="font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none; box-sizing: border-box; display: block; margin-bottom: 10px; padding: 0px; color: rgb(77, 77, 77); font-size: 18px; line-height: 1.5; text-align: left; word-wrap: break-word; font-family: serif; background-color: rgb(255, 255, 255);">For instance, to encrypt the word<span class="Apple-converted-space"> </span><code style="box-sizing: border-box; font-family: &quot;Droid Sans Mono&quot;, monospace; font-weight: inherit; background-color: rgba(0, 0, 0, 0.04); border-radius: 3px; font-size: 15px; line-height: 1.5; padding: 0.1em 0.3em;">“crime”</code></p>

Decrypted message: | c | r | i | m | e
-- | -- | -- | -- | -- | --
Step 1: | 99 | 114 | 105 | 109 | 101
Step 2: | 100 | 214 | 319 | 428 | 529
Step 3: | 100 | 110 | 111 | 116 | 113
Encrypted message: | d | n | o | t | q

<p style="font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none; box-sizing: border-box; display: block; margin-bottom: 10px; padding: 0px; color: rgb(77, 77, 77); font-size: 18px; line-height: 1.5; text-align: left; word-wrap: break-word; font-family: serif; background-color: rgb(255, 255, 255);">The FBI needs an efficient method to decrypt messages. Write a function named<span class="Apple-converted-space"> </span><code style="box-sizing: border-box; font-family: &quot;Droid Sans Mono&quot;, monospace; font-weight: inherit; background-color: rgba(0, 0, 0, 0.04); border-radius: 3px; font-size: 15px; line-height: 1.5; padding: 0.1em 0.3em;">decrypt(word)</code><span class="Apple-converted-space"> </span>that receives a string that consists of small latin letters only, and returns the decrypted word.</p><p style="font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none; box-sizing: border-box; display: block; margin-bottom: 10px; padding: 0px; color: rgb(77, 77, 77); font-size: 18px; line-height: 1.5; text-align: left; word-wrap: break-word; font-family: serif; background-color: rgb(255, 255, 255);">Explain your solution and analyze its time and space complexities.</p><p style="font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none; box-sizing: border-box; display: block; margin-bottom: 10px; padding: 0px; color: rgb(77, 77, 77); font-size: 18px; line-height: 1.5; text-align: left; word-wrap: break-word; font-family: serif; background-color: rgb(255, 255, 255);"><strong style="box-sizing: border-box;">Examples:</strong></p><pre style="font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none; box-sizing: border-box; background-color: rgb(255, 255, 255); border: 1px solid rgb(153, 153, 153); line-height: 1.5; margin: 15px 0px; overflow-x: auto; padding: 0px; caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-size: 18px;"><code class="language-pramp hljs" style="box-sizing: border-box; white-space: pre; background-color: rgba(0, 0, 0, 0.04); color: rgb(56, 58, 66); display: block; overflow-x: auto; padding: 0.1em 0.3em; font-family: &quot;Droid Sans Mono&quot;, monospace; font-weight: 400; border-radius: 3px; font-size: 15px; line-height: 1.5;">input:  word = <span class="hljs-string" style="box-sizing: border-box; color: rgb(80, 161, 79);">"dnotq"</span>
output: <span class="hljs-string" style="box-sizing: border-box; color: rgb(80, 161, 79);">"crime"</span>

input:  word = <span class="hljs-string" style="box-sizing: border-box; color: rgb(80, 161, 79);">"flgxswdliefy"</span>
output: <span class="hljs-string" style="box-sizing: border-box; color: rgb(80, 161, 79);">"encyclopedia"</span>
</code></pre><p style="font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none; box-sizing: border-box; display: block; margin-bottom: 10px; padding: 0px; color: rgb(77, 77, 77); font-size: 18px; line-height: 1.5; text-align: left; word-wrap: break-word; font-family: serif; background-color: rgb(255, 255, 255);">Since the function should be used on messages with many words, make sure the function is as efficient as possible in both time and space. Explain the correctness of your function, and analyze its asymptotic runtime and space complexity.</p><p style="font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none; box-sizing: border-box; display: block; margin-bottom: 10px; padding: 0px; color: rgb(77, 77, 77); font-size: 18px; line-height: 1.5; text-align: left; word-wrap: break-word; font-family: serif; background-color: rgb(255, 255, 255);"><strong style="box-sizing: border-box;">Note:</strong><span class="Apple-converted-space"> </span>Most programing languages have built-in methods of converting letters to ASCII values and vica versa. You may search the internet for the appropriate method.</p><p style="font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none; box-sizing: border-box; display: block; margin-bottom: 10px; padding: 0px; color: rgb(77, 77, 77); font-size: 18px; line-height: 1.5; text-align: left; word-wrap: break-word; font-family: serif; background-color: rgb(255, 255, 255);"><strong style="box-sizing: border-box;">Constraints:</strong></p><ul style="font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none; box-sizing: border-box; margin-bottom: 10px; padding-left: 2em; word-break: break-word; caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-family: serif; font-size: 18px; background-color: rgb(255, 255, 255);"><li style="box-sizing: border-box; color: rgb(77, 77, 77); list-style-type: none; position: relative;"><p style="box-sizing: border-box; display: block; margin-bottom: 10px; padding: 0px; color: rgb(77, 77, 77); font-size: inherit; line-height: 1.5; font-weight: 400; text-align: left; word-wrap: break-word;"><strong style="box-sizing: border-box;">[time limit] 5000ms</strong></p></li><li style="box-sizing: border-box; color: rgb(77, 77, 77); list-style-type: none; position: relative;"><p style="box-sizing: border-box; display: block; margin-bottom: 10px; padding: 0px; color: rgb(77, 77, 77); font-size: inherit; line-height: 1.5; font-weight: 400; text-align: left; word-wrap: break-word;"><strong style="box-sizing: border-box;">[input] string</strong><span class="Apple-converted-space"> </span><code style="box-sizing: border-box; font-family: &quot;Droid Sans Mono&quot;, monospace; font-weight: inherit; background-color: rgba(0, 0, 0, 0.04); border-radius: 3px; font-size: 15px; line-height: 1.5; padding: 0.1em 0.3em;">word</code></p><ul style="box-sizing: border-box; margin-bottom: 10px; padding-left: 2em; word-break: break-word;"><li style="box-sizing: border-box; color: rgb(77, 77, 77); list-style-type: none; position: relative;">The ASCII value of every char is in the range of lowercase letters<span class="Apple-converted-space"> </span><code style="box-sizing: border-box; font-family: &quot;Droid Sans Mono&quot;, monospace; font-weight: inherit; background-color: rgba(0, 0, 0, 0.04); border-radius: 3px; font-size: 15px; line-height: 1.5; padding: 0.1em 0.3em;">a-z</code>.</li></ul></li><li style="box-sizing: border-box; color: rgb(77, 77, 77); list-style-type: none; position: relative;"><p style="box-sizing: border-box; display: block; margin-bottom: 10px; padding: 0px; color: rgb(77, 77, 77); font-size: inherit; line-height: 1.5; font-weight: 400; text-align: left; word-wrap: break-word;"><strong style="box-sizing: border-box;">[output] string</strong></p></li></ul>Decrypt Message

An infamous gang of cyber criminals named “The Gray Cyber Mob”, which is behind many hacking attacks and drug trafficking, has recently become a target for the FBI. After intercepting some of their messages, which looked like complete nonsense, the agency learned that they indeed encrypt their messages, and studied their method of encryption.

Their messages consist of lowercase latin letters only, and every word is encrypted separately as follows:

Convert every letter to its [ASCII value](http://asciitable.com/). Add 1 to the first letter, and then for every letter from the second one to the last one, add the value of the previous letter. Subtract 26 from every letter until it is in the range of lowercase letters a-z in ASCII. Convert the values back to letters.

For instance, to encrypt the word “crime”

Decrypted message:	c	r	i	m	e
Step 1:	99	114	105	109	101
Step 2:	100	214	319	428	529
Step 3:	100	110	111	116	113
Encrypted message:	d	n	o	t	q
The FBI needs an efficient method to decrypt messages. Write a function named decrypt(word) that receives a string that consists of small latin letters only, and returns the decrypted word.

Explain your solution and analyze its time and space complexities.

Examples:

input:  word = "dnotq"
output: "crime"

input:  word = "flgxswdliefy"
output: "encyclopedia"
Since the function should be used on messages with many words, make sure the function is as efficient as possible in both time and space. Explain the correctness of your function, and analyze its asymptotic runtime and space complexity.

Note: Most programing languages have built-in methods of converting letters to ASCII values and vica versa. You may search the internet for the appropriate method.

Constraints:

[time limit] 5000ms

[input] string word

The ASCII value of every char is in the range of lowercase letters a-z.
[output] string

**Comment by esthicodes** [comment]
`#include <iostream>
#include <string>
#include <vector>

using namespace std;

// Dynamic programming O(N) time, O(N) space
int decodeVariations(const string& S) {
	// your code goes here
  int n = S.length();
  vector<int> result(n+1, 0);
  result[n] = 1;
  for (int i = n - 1; i >= 0; i--) {
    if (S[i] == '0') {
      result[i] = 0;
    } else if (S[i] == '1') {
      result[i] = result[i+1] + result[i+2];
    } else if (S[i] == '2') {
      result[i] = result[i+1];
      if (i != n - 1 && S[i+1] < '7') {
        result[i] += result[i+2];
      }
    } else {
      result[i] = result[i+1];
    }
  }
  return result[0];}`

**Reply by PortPolar** [reply]
First of all, notice that the first letter is very easy to decrypt:

Convert the first letter back to its ASCII value.
Subtract 1 from it.
Move the value to be in the range of a-z ASCII values (97-122), by adding 26.
Convert the result back to a character.
The decryption of the rest of the letters is done by almost the same algorithm - given the decrypted previous letter prev, and its value after the second step of encryption - denoted secondStepPrev:

Convert the current letter back to its ASCII value.
Subtract secondStepPrev from it.
Move the value to be in the range of a-z ASCII value (97-122), by adding multiples of 26.
Convert the result back to a character. Store its ASCII value in prev, and add its value to secondStepPrev (for the decryption of the next letter).
Let’s examine the algorithm using the following notation:

dec[n] - the n’th letter before encryption.
enc[n] - the n’th letter after encryption.
secondStep[n] - the n’th letter immediately after step 2 in the encryption.
The encryption algorithm gives the following relation for some integer m (which represents the number of times we need to add 26 to get to an ascii value):

enc[n] = dec[n] + secondStep[n-1] + 26m
By isolating dec[n], we get:

dec[n] = enc[n] - secondStep[n-1] - 26m
Though the value of m isn’t initially known, since the value of the decrypted letter must be in the ASCII range of a-z, the decrypted letter is easily found adding 26’s to enc[n] - secondStep[n] until it is in the right range.

Pseudocode:

function decrypt(word):
    secondStep = 1
    decryption = ""

    for i from 0 to word.length - 1:
        newLetterAscii = asciiValue(word[i])
        newLetterAscii = newLetterAscii - secondStep

        while (newLetterAscii < asciiValue('a')):
            newLetterAscii += 26
         
        decryption = decryption + asciiToChar(newLetterAscii)
        secondStep += newLetterAscii

    return decryption
Note: The following functions were used but not defined, since the implementation is dependant of the programming language:

asciiValue(chr) - returns the ASCII value of a given char.
asciiToChar(chr) - returns the char of a given ASCII value. When your peer programs their solution, it is preferable they use one of the built in functions in the language of their choice, rather than implementing them themselves.
Time Complexity: the function’s asymptotic time complexity is O(N), where N is the length of the input string. the loop that iterates through the letters in the input is performed N times. In the loop, almost every step is done in O(1), except for the loop that is supposed to move the decrypted letter back to the range of a-z. Theoretically, the secondStep may grow linearly with the size of the input. There are two ways to deal with this:

Instead of secondStep itself, we may only keep its remainder after being divided by 26 (since we add/subtract multiples of 26 anyway, the equation dec[N] = enc[N] - (secondStep[N-1] % 26)- 26M still holds, only for a different M). This way all values in every iteration are kept in a constant range.
Note that since in practice this function is used only for words in the English language, the input is bounded and we therefore may ignore the growth of the secondStep anyway.
Space Complexity: the space usage is also O(N) since the output is the same size of the input, and we only keep the output and the second step in storage.

**Comment by PortPolar** [comment]
The most useful way to tackle these kind of problems is using reverse engineering. And for that your peer needs many examples - more than provided.
One way to create examples is by generating them manually. A much more efficient way would be implementing an encryption function (which is straightforward), and using it for testing. Encourage your peer to do so.
The simplest solution to this question is to decrypt one letter at a time, each time using the letters already decrypted. If your peer is stuck, advise them to try finding the relations between the previous decrypted letter and the current encrypted one.
If your peer needs further assistance, try guiding them towards the equation: enc[n] = dec[n] + secondStep[n-1] + 26m, which is fully explained in the solution.
Your peer should be given full score only if they succeed building and testing an efficient decryption function, without any hints. The function should be of O(N) complexity both in terms of time and space.
Make sure your peer gives an accurate complexity analysis, and whether the function they implemented is more efficient when used once on one long word, or many times when used on rather short words.

**Reply by PortPolar** [reply]
The most useful way to tackle these kind of problems is using reverse engineering. And for that your peer needs many examples - more than provided.
One way to create examples is by generating them manually. A much more efficient way would be implementing an encryption function (which is straightforward), and using it for testing. Encourage your peer to do so.
The simplest solution to this question is to decrypt one letter at a time, each time using the letters already decrypted. If your peer is stuck, advise them to try finding the relations between the previous decrypted letter and the current encrypted one.
If your peer needs further assistance, try guiding them towards the equation: enc[n] = dec[n] + secondStep[n-1] + 26m, which is fully explained in the solution.
Your peer should be given full score only if they succeed building and testing an efficient decryption function, without any hints. The function should be of O(N) complexity both in terms of time and space.
Make sure your peer gives an accurate complexity analysis, and whether the function they implemented is more efficient when used once on one long word, or many times when used on rather short words.

**Comment by Prasannapriyan** [comment]
JAVA CODE for Decrypt Message Problem:

import java.io.*;
import java.util.*;

public class DecryptMessage
{
	public static String decrypt(String word) 
	{
		int secondStep = 1;
		int newLetterAscii;
		final StringBuilder decryption = new StringBuilder();

		for (int i = 0; i < word.length(); i++) 
		{
			newLetterAscii = word.charAt(i);
			newLetterAscii -= secondStep;

			while (newLetterAscii < (int) ('a')) 
			{
				newLetterAscii += 26;
			}
			
			decryption.append((char) newLetterAscii);
			secondStep += newLetterAscii;
		}

		return decryption.toString();
	
	}

	public static void main(String[] args) 
    {
        String s = "flgxswdliefy";
        System.out.println("\nGIVEN MESSAGE : " + s);
        System.out.println("\nDECRYPTED MESSAGE : " + decrypt(s));            
    }
}
